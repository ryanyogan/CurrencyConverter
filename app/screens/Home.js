import React, { Component } from 'react';
import { bool, object, func, number, string } from 'prop-types';
import { StatusBar, KeyboardAvoidingView } from 'react-native';
import { connect } from 'react-redux';
import { connectAlert } from '../components/Alert';

import { Container } from '../components/container';
import { Logo } from '../components/Logo';
import { InputWithButton } from '../components/TextInput';
import { ClearButton } from '../components/Buttons';
import { LastConverted } from '../components/Text';
import { Header } from '../components/Header';

import {
  getInitialConverstion,
  swapCurrency,
  changeCurrencyAmount,
} from '../actions/currencies';

class Home extends Component {
  static propTypes = {
    navigation: object, // eslint-disable-line
    dispatch: func,
    baseCurrency: string,
    quoteCurrency: string,
    amount: number,
    conversionRate: number,
    isFetching: bool,
    lastConvertedDate: object, // eslint-disable-line
    primaryColor: string,
    alertWithType: func,
    currencyError: string,
  };

  componentWillMount() {
    this.props.dispatch(getInitialConverstion());
  }

  componentWillReceiveProps(nextProps) {
    if (
      nextProps.currencyError &&
      nextProps.currencyError !== this.props.currencyError
    ) {
      this.props.alertWithType('error', 'Error', nextProps.currencyError);
    }
  }

  handlePressBaseCurrency = () =>
    this.props.navigation.navigate('CurrencyList', {
      title: 'Base Currency',
      type: 'base',
    });

  handlePressQuoteCurrency = () =>
    this.props.navigation.navigate('CurrencyList', {
      title: 'Quote Currency',
      type: 'quote',
    });

  handleTextChange = amount =>
    this.props.dispatch(changeCurrencyAmount(amount));

  handleSwapCurrency = () => this.props.dispatch(swapCurrency());

  handleOptionsPress = () =>
    this.props.navigation.navigate('Options', {
      title: 'Options',
    });

  render() {
    let quotePrice = (this.props.amount * this.props.conversionRate).toFixed(2);
    if (this.props.isFetching) {
      quotePrice = '...';
    }

    return (
      <Container backgroundColor={this.props.primaryColor}>
        <StatusBar translucent={false} barStyle="light-content" />

        <Header onPress={this.handleOptionsPress} />

        <KeyboardAvoidingView behavior="padding">
          <Logo tintColor={this.props.primaryColor} />

          <InputWithButton
            buttonText={this.props.baseCurrency}
            onPress={this.handlePressBaseCurrency}
            onChangeText={this.handleTextChange}
            defaultValue={this.props.amount.toString()}
            keyboardType="numeric"
            textColor={this.props.primaryColor}
          />
          <InputWithButton
            buttonText={this.props.quoteCurrency}
            onPress={this.handlePressQuoteCurrency}
            editable={false}
            value={quotePrice}
            textColor={this.props.primaryColor}
          />
          <LastConverted
            base={this.props.baseCurrency}
            quote={this.props.quoteCurrency}
            conversionRate={this.props.conversionRate}
            date={this.props.lastConvertedDate}
          />
          <ClearButton
            text="Reverse Currencies"
            onPress={this.handleSwapCurrency}
          />
        </KeyboardAvoidingView>
      </Container>
    );
  }
}

const mapStateToProps = state => {
  const baseCurrency = state.currencies.baseCurrency;
  const quoteCurrency = state.currencies.quoteCurrency;
  const conversionSelector = state.currencies.conversions[baseCurrency] || {};
  const rates = conversionSelector.rates || {};

  return {
    baseCurrency,
    quoteCurrency,
    amount: state.currencies.amount,
    conversionRate: rates[quoteCurrency] || 0,
    isFetching: conversionSelector.isFetching,
    lastConvertedDate: conversionSelector.data
      ? new Date(conversionSelector.date)
      : new Date(),
    primaryColor: state.theme.primaryColor,
    currencyError: state.currencies.error,
  };
};

export default connect(mapStateToProps)(connectAlert(Home));
