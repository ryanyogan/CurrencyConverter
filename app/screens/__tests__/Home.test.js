import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import Home from '../Home';

const mockStore = configureStore([]);

it('renders properly', () => {
  const initialState = {
    currencies: {
      baseCurrency: 'USD',
      quoteCurrency: 'GBP',
      conversions: {
        USD: {},
      },
    },
    theme: {
      primaryColor: 'red',
    },
    network: {
      connected: true,
    },
  };
  const store = mockStore(initialState);

  const rendered = shallow(<Home />, {
    context: { store },
  });
  expect(rendered.dive()).toBeTruthy();
  expect(rendered.dive()).toMatchSnapshot();
});
