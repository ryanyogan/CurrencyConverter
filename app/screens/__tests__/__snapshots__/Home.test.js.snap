// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ConnectedAlert
    baseCurrency="USD"
    conversionRate={0}
    conversions={
      Object {
        "USD": Object {},
      }
    }
    currencyError={undefined}
    dispatch={[Function]}
    isConnected={true}
    isFetching={undefined}
    lastConvertedDate={2018-02-06T17:03:44.971Z}
    primaryColor="red"
    quoteCurrency="GBP"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "alert": undefined,
      "alertWithType": undefined,
      "baseCurrency": "USD",
      "conversionRate": 0,
      "conversions": Object {
        "USD": Object {},
      },
      "currencyError": undefined,
      "dispatch": [Function],
      "isConnected": true,
      "isFetching": undefined,
      "lastConvertedDate": 2018-02-06T17:03:44.971Z,
      "primaryColor": "red",
      "quoteCurrency": "GBP",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "alert": undefined,
        "alertWithType": undefined,
        "baseCurrency": "USD",
        "conversionRate": 0,
        "conversions": Object {
          "USD": Object {},
        },
        "currencyError": undefined,
        "dispatch": [Function],
        "isConnected": true,
        "isFetching": undefined,
        "lastConvertedDate": 2018-02-06T17:03:44.971Z,
        "primaryColor": "red",
        "quoteCurrency": "GBP",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
